File: packages/expo-wechat-pay/ios/ExpoWeChatPayModule.swift
import ExpoModulesCore
import WechatOpenSDK

struct WeChatAppPayPayload: Record {
  @Field var appId: String = ""
  @Field var partnerId: String = ""
  @Field var prepayId: String = ""
  @Field var `package`: String = "Sign=WXPay"
  @Field var nonceStr: String = ""
  @Field var timestamp: String = ""
  @Field var sign: String = ""
}

struct WeChatPayOptions: Record {
  @Field var appId: String = ""
  @Field var universalLink: String?
  @Field var payload: WeChatAppPayPayload = WeChatAppPayPayload()
}

public final class ExpoWeChatPayModule: Module {
  static weak var current: ExpoWeChatPayModule?
  private var pendingPayment: Promise?

  public func definition() -> ModuleDefinition {
    Name("ExpoWeChatPay")
    OnCreate { Self.current = self }
    OnDestroy {
      self.rejectPending(code: "ERR_MODULE_DESTROYED", message: "The WeChat Pay module was destroyed.")
      if Self.current === self { Self.current = nil }
    }
    AsyncFunction("isInstalled") { WXApi.isWXAppInstalled() }
    AsyncFunction("pay") { (options: WeChatPayOptions, promise: Promise) in
      guard self.pendingPayment == nil else {
        promise.reject("ERR_PAYMENT_IN_PROGRESS", "Another WeChat payment is already active.")
        return
      }
      guard !options.appId.isEmpty else {
        promise.reject("ERR_INVALID_APP_ID", "A WeChat app ID is required.")
        return
      }
      guard WXApi.registerApp(options.appId, universalLink: options.universalLink ?? "") else {
        promise.reject("ERR_REGISTRATION_FAILED", "The WeChat SDK rejected app registration.")
        return
      }
      guard WXApi.isWXAppInstalled() else {
        promise.reject("ERR_WECHAT_NOT_INSTALLED", "WeChat is not installed.")
        return
      }
      guard let timestamp = UInt32(options.payload.timestamp) else {
        promise.reject("ERR_INVALID_PAYLOAD", "timestamp must be an unsigned integer string.")
        return
      }
      let request = PayReq()
      request.partnerId = options.payload.partnerId
      request.prepayId = options.payload.prepayId
      request.package = options.payload.package
      request.nonceStr = options.payload.nonceStr
      request.timeStamp = timestamp
      request.sign = options.payload.sign
      self.pendingPayment = promise
      WXApi.send(request) { accepted in
        if !accepted { self.rejectPending(code: "ERR_REQUEST_NOT_SENT", message: "WeChat did not accept the payment request.") }
      }
    }
  }

  func completePayment(_ response: PayResp) {
    guard let promise = pendingPayment else { return }
    pendingPayment = nil
    promise.resolve([
      "errorCode": response.errCode,
      "errorMessage": response.errStr as Any,
      "returnKey": response.returnKey as Any,
      "cancelled": response.errCode == -2
    ])
  }

  private func rejectPending(code: String, message: String) {
    guard let promise = pendingPayment else { return }
    pendingPayment = nil
    promise.reject(code, message)
  }
}

File: packages/expo-wechat-pay/android/src/main/java/expo/modules/wechatpay/ExpoWeChatPayModule.kt
package expo.modules.wechatpay

import android.content.Intent
import com.tencent.mm.opensdk.modelbase.BaseReq
import com.tencent.mm.opensdk.modelbase.BaseResp
import com.tencent.mm.opensdk.modelpay.PayReq
import com.tencent.mm.opensdk.modelpay.PayResp
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler
import com.tencent.mm.opensdk.openapi.WXAPIFactory
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class WeChatAppPayPayload : Record {
  @Field val appId: String = ""
  @Field val partnerId: String = ""
  @Field val prepayId: String = ""
  @Field val packageValue: String = "Sign=WXPay"
  @Field val nonceStr: String = ""
  @Field val timestamp: String = ""
  @Field val sign: String = ""
}

class WeChatPayOptions : Record {
  @Field val appId: String = ""
  @Field val universalLink: String? = null
  @Field val payload: WeChatAppPayPayload = WeChatAppPayPayload()
}

class ExpoWeChatPayModule : Module(), IWXAPIEventHandler {
  companion object {
    private var current: ExpoWeChatPayModule? = null
    @JvmStatic fun handleIntent(intent: Intent) { current?.api?.handleIntent(intent, current) }
  }
  private var api: IWXAPI? = null
  private var appId: String? = null
  private var pendingPayment: Promise? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoWeChatPay")
    OnCreate { current = this@ExpoWeChatPayModule }
    OnDestroy {
      rejectPending("ERR_MODULE_DESTROYED", "The WeChat Pay module was destroyed.")
      if (current === this@ExpoWeChatPayModule) current = null
      api = null
    }
    AsyncFunction("isInstalled") {
      val context = appContext.reactContext ?: return@AsyncFunction false
      context.packageManager.getLaunchIntentForPackage("com.tencent.mm") != null
    }
    AsyncFunction("pay") { options: WeChatPayOptions, promise: Promise ->
      if (pendingPayment != null) {
        promise.reject("ERR_PAYMENT_IN_PROGRESS", "Another WeChat payment is already active.", null)
        return@AsyncFunction
      }
      val context = appContext.reactContext
      if (context == null) {
        promise.reject("ERR_NO_ACTIVITY", "The React Native context is unavailable.", null)
        return@AsyncFunction
      }
      if (api == null || appId != options.appId) {
        api = WXAPIFactory.createWXAPI(context, options.appId, true)
        if (api?.registerApp(options.appId) != true) {
          api = null
          promise.reject("ERR_REGISTRATION_FAILED", "The WeChat SDK rejected app registration.", null)
          return@AsyncFunction
        }
        appId = options.appId
      }
      if (api?.isWXAppInstalled != true) {
        promise.reject("ERR_WECHAT_NOT_INSTALLED", "WeChat is not installed.", null)
        return@AsyncFunction
      }
      val payload = options.payload
      val request = PayReq().apply {
        appId = payload.appId
        partnerId = payload.partnerId
        prepayId = payload.prepayId
        packageValue = payload.packageValue
        nonceStr = payload.nonceStr
        timeStamp = payload.timestamp
        sign = payload.sign
      }
      pendingPayment = promise
      if (api?.sendReq(request) != true) rejectPending("ERR_REQUEST_NOT_SENT", "WeChat did not accept the payment request.")
    }
  }
  override fun onReq(request: BaseReq?) = Unit
  override fun onResp(response: BaseResp?) {
    val payment = response as? PayResp ?: return
    val promise = pendingPayment ?: return
    pendingPayment = null
    promise.resolve(mapOf(
      "errorCode" to payment.errCode,
      "errorMessage" to payment.errStr,
      "returnKey" to payment.returnKey,
      "cancelled" to (payment.errCode == BaseResp.ErrCode.ERR_USER_CANCEL)
    ))
  }
  private fun rejectPending(code: String, message: String) {
    val promise = pendingPayment ?: return
    pendingPayment = null
    promise.reject(code, message, null)
  }
}
